File: client/src/models/goal.ts
import { IAccount } from "./account";

export interface IGoalCreateRequest {
  name: string;
  completeDate: Date | null;
  amount: number;
  initialAmount: number | null;
  monthlyContribution: number | null;
  accountIds: string[];
}

export interface IGoalUpdateRequest {
  id: string;
  name: string;
  completeDate: Date | null;
  isCompleteDateEditable: boolean;
  amount: number;
  monthlyContribution: number | null;
  isMonthlyContributionEditable: boolean;
}

export interface IGoalResponse {
  id: string;
  name: string;
  completeDate: Date;
  isCompleteDateEditable: boolean;
  amount: number;
  initialAmount: number;
  monthlyContribution: number;
  isMonthlyContributionEditable: boolean;
  estimatedInterestRate: number | null;
  completed: Date | null;
  accounts: IAccount[];
  userID: string;
}

export enum GoalType {
  None = "",
  SaveGoal = "saveGoal",
  PayGoal = "payGoal",
}

export enum GoalCondition {
  TimedGoal = "timedGoal",
  MonthlyGoal = "monthlyGoal",
}

export enum GoalTarget {
  TargetBalanceGoal = "targetBalanceGoal",
  TargetAmountGoal = "targetAmountGoal",
}

File: server/BudgetBoard.Database/Models/Goal.cs
namespace BudgetBoard.Database.Models;

public class Goal
{
    public Guid ID { get; set; }
    public string Name { get; set; } = string.Empty;
    public DateTime? CompleteDate { get; set; } = null;
    public decimal Amount { get; set; } = 0.0M;
    public decimal InitialAmount { get; set; } = 0.0M;
    public decimal? MonthlyContribution { get; set; } = null;
    public DateTime? Completed { get; set; } = null;
    public ICollection<Account> Accounts { get; set; } = [];
    public required Guid UserID { get; set; }
    public ApplicationUser? User { get; set; } = null;
}
